Patches to the RP4 Build script

ehem opened the pull request:
Things _I_ was noticing during attempts to build.  Due to the way the lampone_build.sh script works, it will retrieve _all_ branches and _all_ tags from your mirrors of the EDK2 repositories.  Due to the size of those repositories those branches mean a fair bit of bandwidth and disk space.  I've also split the repository settings into a distinct file, so if the know okay EDK2 commit changes that is easier.

If one already has a cross-compiling GCC installed, one might want to use the already installed local cross-GCC.  Though now that I think about it, I recall some issues of differing compiler versions causing some issues (erk).

ehem committed:
Direct errors to standard error

Error messages are kind of supposed to go to standard error, not
standard output.

andreiw commented:
Thanks this is good stuff. I'll have to manually merge these, though, since I've made some changes.

ehem committed:
Enable fail on uncaught false return

`set -e` useful for debugging and also indicating if something went
otherwise bonkers.

ehem commented:
Having some interesting times rebasing onto your head and identifying other bits.

`set -e` at times feels a bit hazardous, but the concern is you **really** want the build to fail as early as possible (since that means less waiting for the computer).  Plus you don't want to give the impression it completed successfully when it didn't.

This switches from @ardbiesheuvel's branch of EDK2 to the @pftf branch.  I suspect you would do this yourself in the not too distant future unless there is something I'm unaware of lurking in the pftf branch...

I would _like_ to use /bin/sh instead of /bin/bash (bash is a good interactive shell, but is distinctly slow for scripts; /bin/sh is less featureful, but often significantly faster), but I _suspect_ the edk2/edksetup.sh script **requires** bash (Korne shell is better for advanced scripts).

ehem committed:
Rewrite repository handling

Remove repository configuration from `lampone_build.sh`.  The
configuration is then split into a separate file which can more readily
be changed.

ehem committed:
Commit repository configuration

This is almost as it previously was, but now we're now using the pftf's
fork of the EDK2 repository.

ehem committed:
Adjust checkouts of EDK2 repositories

Part of the build process copies the rebuilt ARM Trusted Firmware over
the versions in edk2-non-osi.  This requires cleaning edk2-non-osi before
doing checkout.

Merge checking out of the branches with the rest of the repository
handling.  Unfortunately this adjustment can only handle the case of two
branches, but that is good enough for now.

Be consistent about checking out remote branches.  If "master" is checked
out a local "master" branch will interfere (though this might be good for
some cases).

ehem commented:
I've been having an interesting time testing, testing and more testing...

Bugs show up and need fixing, so new commits...

ehem commented:
I'm left pondering the ideas of adding an option for doing `git clean -fdx` in the repositories, and adding the option not to update the repositories from GitHub (less network bandwidth, and EDK2 doesn't get that many updates).

andreiw commented:
Thanks for your fixes. I really like what you've done to this. I did a bit of cleanup as it didn't work for me for w/e reason. I think you missed a bit of testing with dirty repos.
